### PR TITLE
[Improvement](statistics) Support for collecting statistics at the granularity of partitions.

### PR DESCRIPTION
**Pull request by weizhengte**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Due to a previous refactoring of the statistics code, this PR mainly enables partitioned-granularity collection of column statistics and column histograms.

Also simplify some code. In addition, to support the subsequent automatic collection of statistics.

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.45 seconds
 stream load tsv:          466 seconds loaded 74807831229 Bytes, about 153 MB/s
 stream load json:         25 seconds loaded 2358488459 Bytes, about 89 MB/s
 stream load orc:          76 seconds loaded 1101869774 Bytes, about 13 MB/s
 stream load parquet:          33 seconds loaded 861443392 Bytes, about 24 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230320131640_clickbench_pr_117453.html

**Comment by weizhengte** [comment]
run buildall

**Comment by weizhengte** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/analysis/AnalyzeStmt.java [hunk 1]
```diff
@@ -166,46 +150,81 @@ private void checkAnalyzePriv(String dbName, String tblName) throws AnalysisExce
...
+            List<String> names = partitionNames.getPartitionNames();
+            Set<String> olapPartitionNames = olapTable.getPartitionNames();
+            List<String> tempPartitionNames = olapTable.getTempPartitions().stream()
+                    .map(Partition::getName).collect(Collectors.toList());
+            Optional<String> optional = names.stream()
+                    .filter(name -> (tempPartitionNames.contains(name)
+                            || !olapPartitionNames.contains(name)))
+                    .findFirst();
+            if (optional.isPresent()) {
+                throw new AnalysisException("Temporary partition or partition does not exist");
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/AnalyzeStmt.java:170]: The exception message seems not match the computation logic  of `optional` variable

File: fe/fe-core/src/main/java/org/apache/doris/statistics/HistogramTask.java [hunk 2]
```diff
@@ -48,7 +49,16 @@ public class HistogramTask extends BaseAnalysisTask {
...
-            + "    `${dbName}`.`${tblName}` TABLESAMPLE (${percentValue} PERCENT)";
+            + "    `${dbName}`.`${tblName}`";
+
+    /** To avoid too much data, use the following efficient sampling method */
+    private static final String ANALYZE_HISTOGRAM_SQL_TEMPLATE_PART = ANALYZE_HISTOGRAM_SQL_TEMPLATE
+            + "    PARTITION (${partName})"
+            + "    TABLESAMPLE (${percentValue} PERCENT)";
+
+    /** To avoid too much data, use the following efficient sampling method */
+    private static final String ANALYZE_HISTOGRAM_SQL_TEMPLATE_FULL = ANALYZE_HISTOGRAM_SQL_TEMPLATE
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/HistogramTask.java:60]: Why full collection also based on table sample function

File: fe/fe-core/src/main/java/org/apache/doris/statistics/MVAnalysisTask.java [hunk 3]
```diff
@@ -111,7 +110,7 @@ public void execute() throws Exception {
                 params.put("idxId", String.valueOf(meta.getIndexId()));
                 String colName = column.getName();
                 params.put("colId", colName);
-                long partId = part.getId();
+                long partId = olapTable.getPartition(partName).getId();
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/MVAnalysisTask.java:113]: olapTable.getPartition(partName) might return null

File: fe/fe-core/src/main/java/org/apache/doris/statistics/AnalysisTaskInfo.java [hunk 4]
```diff
@@ -62,6 +63,8 @@ public enum ScheduleType {
 
     public final String colName;
 
+    public final List<String> partitionNames;
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/statistics/AnalysisTaskInfo.java:66]: Better use `Set` type to deduplicate

File: fe/fe-core/src/main/java/org/apache/doris/analysis/AnalyzeStmt.java [hunk 5]
```diff
@@ -63,48 +61,36 @@ public class AnalyzeStmt extends DdlStmt {
...
-    public boolean isHistogram = false;
-
     private static final ImmutableSet<String> PROPERTIES_SET = new ImmutableSet.Builder<String>()
             .add(CBO_STATISTICS_TASK_TIMEOUT_SEC)
             .build();
 
     private static final Predicate<Long> DESIRED_TASK_TIMEOUT_SEC = (v) -> v > 0L;
 
-    public final boolean wholeTbl;
+    public boolean isWholeTbl;
```
Comment (Kikyou1997) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/AnalyzeStmt.java:70]: Is this variable really necessary?

**Comment by Kikyou1997** [comment]
It'd be much thankful if you would like to add more ut for this

File: fe/fe-core/src/main/java/org/apache/doris/analysis/AnalyzeStmt.java (hunk 5 above)
Comment (weizhengte) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/AnalyzeStmt.java:70]: This is an existing parameter used by`MVAnalysisTask` to make logical judgment:  ”Analysis for the materialized  view, only gets constructed when the AnalyzeStmt is not set which columns to be analyzed“. It doesn't seem to need to be removed

File: fe/fe-core/src/main/java/org/apache/doris/statistics/AnalysisTaskInfo.java (hunk 4 above)
Comment (weizhengte) [reply fe/fe-core/src/main/java/org/apache/doris/statistics/AnalysisTaskInfo.java:66]: Sure

File: fe/fe-core/src/main/java/org/apache/doris/statistics/HistogramTask.java (hunk 2 above)
Comment (weizhengte) [reply fe/fe-core/src/main/java/org/apache/doris/statistics/HistogramTask.java]: yeah, this should be in the sample task(Next pr)

File: fe/fe-core/src/main/java/org/apache/doris/statistics/MVAnalysisTask.java (hunk 3 above)
Comment (weizhengte) [reply fe/fe-core/src/main/java/org/apache/doris/statistics/MVAnalysisTask.java]: Fixed

**Comment by weizhengte** [comment]
> It'd be much thankful if you would like to add more ut for this

Okay

**Comment by weizhengte** [comment]
run buildall

**Review by Kikyou1997 — APPROVED** [review]

**Review by englefly — APPROVED** [review]
LGTM

**Pull request merged**